torch/functional.py

```diff
@@ -1589,3 +1591,54 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
 
 def align_tensors(*tensors):
     raise RuntimeError('`align_tensors` not yet implemented.')
+
+def result_type(*arrays_and_dtypes):
+    """result_type(*arrays_and_dtypes: Union[Tensor, dtype, Number]) -> dtype
+
+Returns the :class:`torch.dtype` that would result from performing an arithmetic
+operation on the provided input tensors. See type promotion :ref:`documentation <type-promotion-doc>`
+for more information on the type promotion logic.
+
+Args:
+    arrays_and_dtypes (Tensor, Number or dtype): input tensors, numbers or dtypes
+
+Example::
+
+    >>> torch.result_type(torch.tensor([1, 2], dtype=torch.int), 1.0)
+    torch.float32
+    >>> torch.result_type(torch.tensor([1, 2], dtype=torch.uint8), torch.tensor(1))
+    torch.uint8
+    >>> torch.result_type(torch.int32, torch.float32)
+    torch.float32
+    >>> torch.result_type(torch.tensor([1, 2], dtype=torch.int8), 1.0, torch.float32)
+    torch.float32
+"""
+
+    tensors: List[torch.Tensor] = []
```

Comment:
Possible alternative implementation of this, inspired by your proposed reference implementation above:

```
assert len(arrays_and_dtypes) > 0

tensors = [t for t in arrays_and_dtypes if isinstance(t, torch.Tensor)]
scalars = [s for s in arrays_and_dtypes if isinstance(t, Number)]
dtypes = [d for d in arrays_and_dtypes if isinstance(t, torch.dtype)]

tensor_dtype = None if len(tensors) == 0 else _VF._result_type(tensors)
scalar_dtype = None if len(scalars) == 0 else _VF._result_type_scalars(scalars)
dtype_dtype = None if len(dtypes) == 0 else _VF._result_type_dtypes(dtypes)

return (torch.tensor([], dtype=tensor_dtype) + torch.tensor([], dtype=dtype_dtype) + torch.tensor(0, dtype=scalar_dtype)).dtype
```

Follow-up question: will this or an implementation like the current version work with the jit scripting and tracing? Is there a test for that?

If the jit is still an issue let me know ASAP and I can help modify the PR to fix that.

Reply:
I have confirmed that the jit scripting/tracing test works fine locally, but I have not added tests yet. In which file should I add the tests?